source/extensions/http/cache/file_system_http_cache/file_system_http_cache.cc

```diff
@@ -354,14 +366,15 @@ void FileSystemHttpCache::init() {
   if (config().has_max_cache_entry_count()) {
     stats_.size_limit_count_.set(config().max_cache_entry_count().value());
   }
+  cache_eviction_thread_.addCache(*this);
 }
 
 void FileSystemHttpCache::trackFileAdded(uint64_t file_size) {
   size_count_++;
   size_bytes_ += file_size;
   stats_.size_count_.inc();
   stats_.size_bytes_.add(file_size);
-  // TODO(ravenblack): potentially trigger cache cleanup operation.
+  cache_eviction_thread_.signal();
```

Comment:
I'd suggest keeping an approximate track of the size of the cache overall and only signaling the eviction thread when either:
 * the approximate size (ignoring other processes sharing the cache) exceeds a threshold
 * a timer we can tune goes off; which would do eventual cleanup if there were two overlapping Envoy processes sharing the cache config, e.g. for hot-restart or because someone wants to run more than one distinct Envoy server from the same machine.

My concern is that every new file that gets added will trigger a full sweep of the disk, which will impact system performance. I think it'd be good to have in mind an eviction strategy where we do evictions when the file system reaches X bytes, and bring it down to 0.75*X bytes, or similar, so that we aren't thrashing the disk in the extra thread.

I think it's OK to have an extra thread doing the GC but I'd be concerned if that was constantly busy.

Reply:
Added a comment here to clarify that signaling does not start an eviction, only a *maybe* eviction - the eviction thread only does any disk-work if the current cache size exceeds a threshold.

I do have in mind softening options because as you say even with a threshold check, it still might be constantly sweeping if the cache keeps repopulating. Either high/low thresholds (erase to low when it hits high, like watermarks) or a threshold and a time limit (erase only if threshold exceeded *and* it's been n seconds since last erase), but this impl is MVP.

Reply:
The train of thought was that we generally want envoy threads freed as quickly as possible, and signal-always is probably cheaper *in that thread* than touching two atomics. And then the eviction thread won't actually wake up every time, because (assuming things are busy) many times it will already be awake and the signal won't do anything. But you're probably right that most of the time we should be doing nothing, and touching two atomics is cheap.

On the downside I ended up with two versions of checking if we need to evict, because one place we need to know *how much* to evict, which is more work, so we probably don't want to do that in the envoy worker thread unnecessarily.